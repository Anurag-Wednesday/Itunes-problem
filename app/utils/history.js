import { createBrowserHistory } from 'history';
import routeConstants from '@utils/routeConstants';

export const setbaseUrl = () => {
  let baseUrl = '';
  const routes = Object.keys(routeConstants);
  const pathname = window.location.pathname;
  if (process.env.ENVIRONMENT_NAME === 'uat') {
    routes.forEach((routeKey) => {
      let route = routeConstants[routeKey].route;
      const pathnames = pathname.split('/');
      const ids = [];
      let index = 0;
      pathnames.forEach((p) => {
        const num = parseInt(p, 10);
        if (!Number.isNaN(num)) {
          ids.push({ value: p, startIndex: index });
        }

        index += p.length + 1;
      });
      let idCounter = 0;
      while (ids.length && idCounter < ids.length && route.includes(':')) {
        const currentSegmentId = pathname.substring(
          ids[idCounter].startIndex,
          ids[idCounter].startIndex + ids[idCounter].value.length
        );
        const lastIndexOfColon = route.indexOf(':');
        let indexOfSlash = route.length;
        for (let i = lastIndexOfColon; i < route.length; i++) {
          if (route.charAt(i) === '/') {
            indexOfSlash = i;
            break;
          }
        }
        route = route.replace(route.substring(lastIndexOfColon, indexOfSlash), currentSegmentId);
        idCounter++;
      }
      if (pathname.includes(route)) {
        if (pathname.substring(pathname.length - route.length, pathname.length) === route) {
          baseUrl = pathname.substring(pathname.length - route.length);
        }
        if (pathname.substring(pathname.length - route.length - 1, pathname.length) === `${route}/`) {
          baseUrl = pathname.substring(0, pathname.length - route.length - 1);
        }
      }
    });
  }
  return baseUrl;
};

const history = createBrowserHistory(setbaseUrl());
export default history;
