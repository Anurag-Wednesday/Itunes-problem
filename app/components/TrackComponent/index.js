import React, { memo, useState } from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';
import { Card, Modal } from 'antd';
import { colors } from '@app/themes/index';
import styled from 'styled-components';
import T from '@components/T';
import If from '@components/If';
import history from '@app/utils/history';

const TrackCard = styled(Card)`
  && {
    width: 22rem;
    height: auto;
    alignitems: start;
    flexwrap: wrap;
    background-color: ${colors.background};
    color: white;
    font-weight: 400;
    padding: 1rem;
  }
`;

const TrackImage = styled.img`
  && {
    width: 6.25rem;
    height: 6.25rem;
    margin-left: 30%;
    margin-bottom: 1rem;
    border-radius: 50%;
  }
`;

const AudioPlayer = styled.audio`
  && {
    width: 15.6rem;
    height: 2.2rem;
    margin-top: 1.5rem;
  }
`;

const Title = styled(T)`
  && {
    color: ${colors.whiteText};
    font-style: italic;
  }
`;

const StyledText = styled(T)`
  && {
    color: ${colors.text};
  }
`;
export function TrackComponent({ trackId, trackName, artworkUrl100, artistName, previewUrl, handleOnActionClick }) {
  const [visible, setVisible] = useState(false);
  function onActionHandler(evt) {
    handleOnActionClick(evt.target, evt.type);
  }

  return (
    <>
      <TrackCard
        onClick={() => setVisible(true)}
        data-testid="TrackCard"
        title={
          <If condition={!isEmpty(trackName)} otherwise={<T type="subheading" id="no_track_name" />}>
            <Title type="subheading" id="name" values={{ trackName }} />
          </If>
        }
      >
        <If condition={!isEmpty(artworkUrl100)} otherwise={<T id="image_unavailable" />}>
          <TrackImage alt={trackName} src={artworkUrl100} />
        </If>
        <If condition={!isEmpty(artistName)} otherwise={<T id="artist_data_unavailable" />}>
          <T type="subText" data-testid="artist_name" id="artist_name" values={{ artist: artistName }} />
        </If>
        <If condition={!isEmpty(trackName)} otherwise={trackName}>
          <T type="subText" data-testid="name" id="name" values={{ trackName }} />
        </If>
        <If condition={!isEmpty(previewUrl)} otherwise={<T id="preview_unavailable" />}>
          <AudioPlayer
            data-testid={trackName}
            id={trackName}
            onPlay={(evt) => onActionHandler(evt)}
            onPause={(evt) => onActionHandler(evt)}
            controls
          >
            <source src={previewUrl} />
          </AudioPlayer>
        </If>
      </TrackCard>
      <Modal
        data-testid="modal"
        title="Leave Page?"
        visible={visible}
        onCancel={() => setVisible(false)}
        onOk={() => history.push(`/tracks/${trackId}`)}
      >
        <StyledText type="subText" id="modal_text" values={{ track: trackName }} />
      </Modal>
    </>
  );
}

TrackComponent.propTypes = {
  trackId: PropTypes.number,
  trackName: PropTypes.string,
  previewUrl: PropTypes.string,
  artistName: PropTypes.string,
  artworkUrl100: PropTypes.string,
  handleOnActionClick: PropTypes.func
};

export default memo(TrackComponent);
