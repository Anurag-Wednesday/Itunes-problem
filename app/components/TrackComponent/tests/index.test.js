import React from 'react';
import { Router } from 'react-router';
import { renderWithIntl, timeout } from '@utils/testUtils';
import { fireEvent } from '@testing-library/dom';
import history from '@app/utils/history';
import TrackComponent from '../index';
import { translate } from '@app/components/IntlGlobalProvider/index';

describe('<TrackComponent />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TrackComponent />);
    expect(baseElement).toMatchSnapshot();
  });
});

it('should contain one TrackComponent', () => {
  const { getAllByTestId } = renderWithIntl(<TrackComponent />);
  expect(getAllByTestId('TrackCard').length).toBe(1);
});

it('should render the track details inside the Card', () => {
  const trackName = 'See you Again';
  const artistName = 'Charlie Puth';
  const { getByTestId } = renderWithIntl(<TrackComponent trackName={trackName} artistName={artistName} />);
  expect(getByTestId('name')).toHaveTextContent(trackName);
  expect(getByTestId('artist_name')).toHaveTextContent(artistName);
});

it('should call onActionHandler if audio controls are clicked', () => {
  const handleOnActionClick = jest.fn();
  const trackName = 'See you Again';
  const artistName = 'Charlie Puth';
  const previewUrl =
    'https://is4-ssl.mzstatic.com/image/thumb/Podcasts125/v4/72/46/63/724663b9-46ac-ab25-c167-546ef48f7ed5/mza_1727273594955964910.jpg/60x60bb.jpg';
  const artworkUrl100 =
    'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/5d/b0/e4/5db0e413-9d75-d04c-5e95-ea9fc6361084/mzaf_16693952033856250131.plus.aac.p.m4a';
  const { getByTestId } = renderWithIntl(
    <TrackComponent
      trackName={trackName}
      artistName={artistName}
      previewUrl={previewUrl}
      artworkUrl100={artworkUrl100}
      handleOnActionClick={handleOnActionClick}
    />
  );
  fireEvent.pause(getByTestId(trackName));
  expect(handleOnActionClick).toBeCalledWith(getByTestId(trackName), 'pause');
  expect(getByTestId(trackName).paused).toEqual(true);
  fireEvent.play(getByTestId(trackName));
  expect(handleOnActionClick).toBeCalledWith(getByTestId(trackName), 'play');
});

it('should render the track details unavailable messages in case any props are unavailable', () => {
  const { getByText } = renderWithIntl(<TrackComponent />);
  const trackNameUnavailable = translate('no_track_name');
  const artistDataUnavailable = translate('artist_data_unavailable');
  const imageUnavailable = translate('image_unavailable');
  const previewUnavailable = translate('preview_unavailable');

  expect(getByText(trackNameUnavailable)).toBeInTheDocument();
  expect(getByText(artistDataUnavailable)).toBeInTheDocument();
  expect(getByText(imageUnavailable)).toBeInTheDocument();
  expect(getByText(previewUnavailable)).toBeInTheDocument();
});

it('should redirect to /trackId when clicked on Clickable component', async () => {
  const trackName = 'See you Again';
  const artistName = 'Charlie Puth';
  const trackId = 1445140962;
  const previewUrl =
    'https://is4-ssl.mzstatic.com/image/thumb/Podcasts125/v4/72/46/63/724663b9-46ac-ab25-c167-546ef48f7ed5/mza_1727273594955964910.jpg/60x60bb.jpg';
  const artworkUrl100 =
    'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/5d/b0/e4/5db0e413-9d75-d04c-5e95-ea9fc6361084/mzaf_16693952033856250131.plus.aac.p.m4a';
  const { getByTestId } = renderWithIntl(
    <Router history={history}>
      <TrackComponent
        trackName={trackName}
        artistName={artistName}
        previewUrl={previewUrl}
        artworkUrl100={artworkUrl100}
        trackId={trackId}
      />
    </Router>
  );
  const historySpy = jest.spyOn(history, 'push');
  fireEvent.click(getByTestId('TrackCard'));
  await timeout(500);
  expect(historySpy).toHaveBeenCalledWith(`/tracks/${trackId}`);
});
