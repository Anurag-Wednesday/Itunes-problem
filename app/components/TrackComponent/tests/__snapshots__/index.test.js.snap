// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackComponent /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered TrackComponent__TrackCard-sc-1y7fh5t-0 cFlwQU"
      data-testid="TrackCard"
    >
      <div
        class="ant-card-head"
      >
        <div
          class="ant-card-head-wrapper"
        >
          <div
            class="ant-card-head-title"
          >
            <p
              class="T__StyledText-gjlic1-0 iGfozg"
              data-testid="t"
            >
              No track name available
            </p>
          </div>
        </div>
      </div>
      <div
        class="ant-card-body"
      >
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="t"
        >
          Image Unavailable
        </p>
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="t"
        >
          No artist data available
        </p>
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="t"
        >
          No preview available
        </p>
      </div>
    </div>
  </div>
</body>
`;
